### PR TITLE
Enable error linter

### PR DESCRIPTION
**Pull request by nojaf**

### CODE REVIEW

File: mlflow/go/README.md [hunk 1]
```diff
@@ -74,3 +74,11 @@ See the [pytest documentation](https://docs.pytest.org/en/8.2.x/how-to/usage.htm
 ## Supported endpoints
 
 The currently supported endpoints can be found in [mlflow/tools/generate/endspoints.go](./tools/generate/endspoints.go).
+
+## Linters
+
+We have enabled various linters from [golangci-lint](https://golangci-lint.run/), you can run these via:
+
+```bash
+golangci-lint run ./mlflow/go/...
```
Comment (jgiannuzzi) [inline mlflow/go/README.md:83]:
```suggestion
golangci-lint run
```

File: mlflow/go/pkg/contract/error.go [hunk 2]
```diff
@@ -15,6 +15,7 @@ func (e ErrorCode) String() string {
 
 // Custom json marshalling for ErrorCode.
 func (e ErrorCode) MarshalJSON() ([]byte, error) {
+	//nolint:all
```
Comment (jgiannuzzi) [inline mlflow/go/pkg/contract/error.go:18]: shouldn't we specify exactly which linters we want to disable?

File: mlflow/go/pkg/query/parser/validate.go [hunk 3]
```diff
@@ -90,7 +92,7 @@ func parseValidIdentifier(identifier string) (ValidIdentifier, error) {
 	case datasetIdentifier, "datasets":
 		return Dataset, nil
 	default:
-		return -1, fmt.Errorf("invalid identifier: %s", identifier)
+		return -1, fmt.Errorf("invalid identifier \"%s\": %w", identifier, ErrValidation)
```
Comment (jgiannuzzi) [inline mlflow/go/pkg/query/parser/validate.go:95]: use `%q` if you want to quote the string - also worth checking other places where we'd want to quote for consistency

File: mlflow/go/pkg/server/command.go [hunk 4]
```diff
@@ -32,8 +33,9 @@ func launchCommand(ctx context.Context, cfg *config.Config) error {
 
 	logrus.Debugf("Launching command: %v", cmd)
 	if err := cmd.Start(); err != nil {
-		return err
+		return fmt.Errorf("command could not launch: %w", err)
 	}
 
-	return cmd.Wait()
+	err := cmd.Wait()
+	return fmt.Errorf("command exited with error: %w", err)
```
Comment (jgiannuzzi) [inline mlflow/go/pkg/server/command.go:40]:
```suggestion
	if err != nil {
		return fmt.Errorf("command exited with error: %w", err)
	}

	return nil
```

File: mlflow/go/pkg/server/server.go [hunk 5]
```diff
@@ -74,13 +75,14 @@ func launchServer(ctx context.Context, cfg *config.Config) error {
 			break
 		}
 		if errors.Is(err, context.Canceled) {
-			return err
+			return fmt.Errorf("could not dial Python server: %w", err)
```
Comment (jgiannuzzi) [inline mlflow/go/pkg/server/server.go:78]:
```suggestion
			return fmt.Errorf("could not connect to Python server: %w", err)
```

File: mlflow/go/pkg/server/server.go [hunk 6]
```diff
@@ -74,13 +75,14 @@ func launchServer(ctx context.Context, cfg *config.Config) error {
...
 		}
 		time.Sleep(1 * time.Second)
 	}
 	logrus.Debugf("Python server is ready")
 
-	return app.Listen(cfg.Address)
+	err = app.Listen(cfg.Address)
+	return fmt.Errorf("failed to start MLflow experimental Go server: %w", err)
```
Comment (jgiannuzzi) [inline mlflow/go/pkg/server/server.go:85]: do the `if err != nil` thing please

File: mlflow/go/pkg/server/validation.go [hunk 7]
```diff
@@ -33,7 +34,7 @@ func NewValidator() (*validator.Validate, error) {
 			return value > -1
 		},
 	); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("validation registration for 'stringAsPositiveInteger' failed %w", err)
```
Comment (jgiannuzzi) [inline mlflow/go/pkg/server/validation.go:37]:
```suggestion
		return nil, fmt.Errorf("validation registration for 'stringAsPositiveInteger' failed: %w", err)
```

File: mlflow/go/pkg/server/validation.go [hunk 8]
```diff
@@ -53,7 +54,7 @@ func NewValidator() (*validator.Validate, error) {
 			return u.Fragment == "" && u.RawQuery == "" && !strings.Contains(u.RawQuery, "..")
 		},
 	); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("validation registration for 'uriWithoutFragmentsOrParamsOrDotDotInQuery' failed %w", err)
```
Comment (jgiannuzzi) [inline mlflow/go/pkg/server/validation.go:57]:
```suggestion
		return nil, fmt.Errorf("validation registration for 'uriWithoutFragmentsOrParamsOrDotDotInQuery' failed: %w", err)
```

File: mlflow/go/pkg/query/parser/parser.go [hunk 9]
```diff
@@ -41,9 +42,11 @@ func (p *parser) advance() lexer.Token {
 	return tk
 }
 
+var ErrParserError = errors.New("parser error")
+
 func (p *parser) parseIdentifier() (Identifier, error) {
 	if p.hasTokens() && p.currentTokenKind() != lexer.Identifier {
-		return Identifier{}, fmt.Errorf("Expected identifier, got %s", p.printCurrentToken())
+		return Identifier{}, fmt.Errorf("expected identifier, got %s: %w", p.printCurrentToken(), ErrParserError)
```
Comment (jgiannuzzi) [inline mlflow/go/pkg/query/parser/parser.go:49]: I'm not too sure about how we now wrap our own errors in a new one with some text. I think that instead we should define our own error struct with an `Error()` method, and that returns the custom text. Let me know if this is unclear!

Comment (nojaf) [reply mlflow/go/pkg/query/parser/parser.go:49]: Please give an example to clarify.

File: mlflow/go/README.md (hunk 1 above)
Comment (nojaf) [reply mlflow/go/README.md:83]: I slightly disagree here, is letting it loose on the root folder not going to try and find `*.go` files in all folders? Perhaps not noticeable today but you should really only lint our own Go code.

Comment (jgiannuzzi) [reply mlflow/go/README.md:83]: And also `pre-commit run golangci-lint --all-files` to run on all the files, and not just the staged ones.

Comment (nojaf) [reply mlflow/go/README.md:83]: I'm just somewhat worried that some unexpected folders will at some point contain Go code.
For example, does it skip `node_modules` under `mlflow/server/js`?

Comment (jgiannuzzi) [reply mlflow/go/README.md:83]: that's something we should then configure in `.pre-commit-config.yaml`, as the CI _will_ fail otherwise

**Pull request merged**